mobile/library/common/extensions/filters/http/network_configuration/filter.cc

```diff
@@ -55,6 +60,33 @@ bool NetworkConfigurationFilter::onAddressResolved(
   return false;
 }
 
+void NetworkConfigurationFilter::onProxyResolutionComplete(
+    Network::ProxySettingsConstSharedPtr proxy_settings) {
+  if (continueWithProxySettings(proxy_settings) == Http::FilterHeadersStatus::Continue) {
+    continue_decoding_callback_ = decoder_callbacks_->dispatcher().createSchedulableCallback(
+        [this]() { decoder_callbacks_->continueDecoding(); });
+    continue_decoding_callback_->scheduleCallbackNextIteration();
+  }
+}
+
+void proxyResolutionCompleted(envoy_proxy_settings_list proxy_settings_list, const void* context) {
+  const auto wrapped_filter =
+      const_cast<std::unique_ptr<std::weak_ptr<NetworkConfigurationFilter>>*>(
+          static_cast<const std::unique_ptr<std::weak_ptr<NetworkConfigurationFilter>>*>(context));
+
+  // Release proxy settings list no matter whether filter still exists or not.
+  const auto proxy_settings = Network::ProxySettings::create(proxy_settings_list);
+  envoy_proxy_settings_list_release(proxy_settings_list);
+
+  const auto filter = wrapped_filter->get()->lock();
```

Comment:
can we get comments somewhere on how the lifetime cycle works for this? I assume this is handling the case that the stream is canceled during async resolution but ideally we'd have this documented somewhere (maybe it is and I just haven't found it =P)

Reply:
I reworked this so we don't use C malloc and instead rely on C++ smart pointers.  However, registration of the API (i.e. registerApi) requires passing in a raw pointer for now until we rework that, so I commented on the lifetime and ownership there.